src/pip/_internal/wheel.py

```diff
@@ -1036,19 +1037,25 @@ def build(
         # type: (...) -> List[InstallRequirement]
         """Build wheels.
 
-        :param unpack: If True, replace the sdist we built from with the
-            newly built wheel, in preparation for installation.
+        :param autobuilding: If True, replace the sdist we built from with the
+            unpacked newly built wheel, in preparation for installation.
         :return: True if all the wheels built correctly.
         """
         buildset = []
         format_control = self.finder.format_control
-        # Whether a cache directory is available for autobuilding=True.
-        cache_available = bool(self._wheel_dir or self.wheel_cache.cache_dir)
+        # pip install uses autobuilding=True.
+        # pip install never provides a _wheel_dir.
+        # pip wheel uses autobuilding=False.
+        # pip wheel always provides a _wheel_dir (via the preparer).
+        assert (
+            (autobuilding and not self._wheel_dir) or
+            (not autobuilding and self._wheel_dir)
+        )
 
         for req in requirements:
             ephem_cache = should_use_ephemeral_cache(
                 req, format_control=format_control, autobuilding=autobuilding,
-                cache_available=cache_available,
+                cache_available=bool(self.wheel_cache.cache_dir),
```

Comment:
Are we sure the wheel cache is available at all call sites?

Reply:
@pradyunsg goog catch. A unit test caught it too. Fixed.